### ISSUE TITLE
Cannot load related collections for transient objects, e.g. packages do not explode

### ISSUE THREAD

**Issue by kbower** [kbower/tickettest1#1349]
_Imported from trac ticket 1349.
Created by **kbower**
Opened in trac: Jan 9, 2012, 12:25:57 PM
Last modified in trac: Jan 10, 2012, 2:12:21 PM_
***
sqlalchemy upgrade issue

see http://groups.google.com/group/sqlalchemy/browse_thread/thread/e9b99d7120e63591
and http://www.sqlalchemy.org/trac/changeset/d6e321dc120d/

**Comment by kbower** [comment]
_**kbower** trac comment on Jan 10, 2012, 2:12:21 PM:_

(Changeset [[2131]](../commit/c509bf56a588724b621c1127c838e46f95092050))
 * apply patch to sqlalchemy transient collection loading (fixes #1349)
   * see http://groups.google.com/group/sqlalchemy/browse_thread/thread/e9b99d7120e63591
   * see http://www.sqlalchemy.org/trac/changeset/d6e321dc120d/
```diff
diff -rU10 SQLAlchemy-0.7.4.90kb/lib/sqlalchemy/__init__.py SQLAlchemy-0.7.4.91kb/lib/sqlalchemy/__init__.py
--- SQLAlchemy-0.7.4.90kb/lib/sqlalchemy/__init__.py    2012-01-04 17:34:33.000000000 -0500
+++ SQLAlchemy-0.7.4.91kb/lib/sqlalchemy/__init__.py    2012-01-10 13:58:29.000000000 -0500
@@ -110,16 +110,16 @@
     ThreadLocalMetaData,
     UniqueConstraint,
     )
 
 from sqlalchemy.engine import create_engine, engine_from_config
 
 
 __all__ = sorted(name for name, obj in locals().items()
                  if not (name.startswith('_') or inspect.ismodule(obj)))
 
-__version__ = '0.7.4.90kb'
+__version__ = '0.7.4.91kb'
 
 del inspect, sys
 
 from sqlalchemy import util as _sa_util
 _sa_util.importlater.resolve_all()
diff -rU10 SQLAlchemy-0.7.4.90kb/lib/sqlalchemy/sql/compiler.py SQLAlchemy-0.7.4.91kb/lib/sqlalchemy/sql/compiler.py
--- SQLAlchemy-0.7.4.90kb/lib/sqlalchemy/sql/compiler.py    2011-12-15 11:42:50.000000000 -0500
+++ SQLAlchemy-0.7.4.91kb/lib/sqlalchemy/sql/compiler.py    2012-01-10 13:57:14.000000000 -0500
@@ -298,32 +298,27 @@
                 elif bindparam.required:
                     if _group_number:
                         raise exc.InvalidRequestError(
                                 "A value is required for bind parameter %r, "
                                 "in parameter group %d" % 
                                 (bindparam.key, _group_number))
                     else:
                         raise exc.InvalidRequestError(
                                 "A value is required for bind parameter %r" 
                                 % bindparam.key)
-                elif bindparam.callable:
-                    pd[name] = bindparam.callable()
                 else:
-                    pd[name] = bindparam.value
+                    pd[name] = bindparam.effective_value
             return pd
         else:
             pd = {}
             for bindparam in self.bind_names:
-                if bindparam.callable:
-                    pd[self.bind_names[bindparam]] = bindparam.callable()
-                else:
-                    pd[self.bind_names[bindparam]] = bindparam.value
+                pd[self.bind_names[bindparam]] = bindparam.effective_value
             return pd
 
     params = property(construct_params, doc="""
         Return the bind params for this compiled object.
 
     """)
 
     def default_from(self):
         """Called when a SELECT statement has no froms, and no FROM clause is
         to be appended.
diff -rU10 SQLAlchemy-0.7.4.90kb/lib/sqlalchemy/sql/expression.py SQLAlchemy-0.7.4.91kb/lib/sqlalchemy/sql/expression.py
--- SQLAlchemy-0.7.4.90kb/lib/sqlalchemy/sql/expression.py  2011-12-15 11:42:50.000000000 -0500
+++ SQLAlchemy-0.7.4.91kb/lib/sqlalchemy/sql/expression.py  2012-01-10 13:57:14.000000000 -0500
@@ -2597,20 +2597,35 @@
                     _compared_to_type._coerce_compared_value(
                         _compared_to_operator, value)
             else:
                 self.type = sqltypes._type_map.get(type(value),
                         sqltypes.NULLTYPE)
         elif isinstance(type_, type):
             self.type = type_()
         else:
             self.type = type_
 
+    @property
+    def effective_value(self):
+        """Return the value of this bound parameter, 
+        taking into account if the ``callable`` parameter
+        was set.  
+        
+        The ``callable`` value will be evaluated
+        and returned if present, else ``value``.
+        
+        """
+        if self.callable:
+            return self.callable()
+        else:
+            return self.value
+
     def _clone(self):
         c = ClauseElement._clone(self)
         if self.unique:
             c.key = _generated_label('%%(%d %s)s' % (id(c), c._orig_key
                     or 'param'))
         return c
 
     def _convert_to_unique(self):
         if not self.unique:
             self.unique = True
diff -rU10 SQLAlchemy-0.7.4.90kb/lib/sqlalchemy/sql/util.py SQLAlchemy-0.7.4.91kb/lib/sqlalchemy/sql/util.py
--- SQLAlchemy-0.7.4.90kb/lib/sqlalchemy/sql/util.py    2011-12-15 11:42:50.000000000 -0500
+++ SQLAlchemy-0.7.4.91kb/lib/sqlalchemy/sql/util.py    2012-01-10 13:57:14.000000000 -0500
@@ -145,27 +145,21 @@
 
         >>> expr = and_(
         ...    table.c.foo==5, table.c.foo==7
         ... )
         >>> bind_values(expr)
         [5, 7]
     """
 
     v = []
     def visit_bindparam(bind):
-        value = bind.value
-
-        # evaluate callables
-        if callable(value):
-            value = value()
-
-        v.append(value)
+        v.append(bind.effective_value)
 
     visitors.traverse(clause, {}, {'bindparam':visit_bindparam})
     return v
 
 def _quote_ddl_expr(element):
     if isinstance(element, basestring):
         element = element.replace("'", "''")
         return "'%s'" % element
     else:
         return repr(element)
```

**Comment by kbower** [comment]
_**kbower** trac status change on Jan 10, 2012, 2:12:21 PM:_ **closed**
